### PR TITLE
Use fulfilment date calculator for start date

### PR DESCRIPTION
**Pull request by frj**
These changes modify the new product api endpoint to make use of the [fulfilment date calculator](https://github.com/guardian/support-service-lambdas/tree/master/handlers/fulfilment-date-calculator) to calculate the first available date for a subscription.

This fixes a [bug](https://trello.com/c/ZusfbzZo/1124-add-validation-to-hd-delivery-dates-migrate-new-product-api-to-use-the-fulfilment-date-calculator) which allowed CSRs to create home delivery subscriptions starting on dates before they could actually be fulfilled.

The changes add an addition field to the catalog endpoint which salesforce uses to build the various product/rateplan/startdate drop downs in the Billing Account -> Create Subscription page.

```json
{
  "products": [
    {
      "plans": [
        {
            "selectableWindow": {
                "startDate": "2020-04-29",
                ...
```

Details:
- Fulfilment date files are downloaded from s3
- Start dates for each rate plan are calculated by filtering the dates from the fulfilment date file based on the ProductType and the Days of the week relevant to the particular rate plan. Then taking the soonest of those dates as the fist available date for that rateplan. 
- The calculated start date is exposed in the get-cataloge endpoint so salesforce can display the correct date. The cutOffDayInclusive and startDaysAfterCutOff are now redundant will be removed once salesforce has been migrated use the new field.
- The start date is used in the validation of the start date supplied in  add-subscription endpoint

### CODE REVIEW

**Comment by frj** [comment]
[Add validation to HD Delivery Dates: update new product api catalog endpoint to use fulfilment date calculator](https://trello.com/c/Il3hBF8T/1242-add-validation-to-hd-delivery-dates-update-new-product-api-catalog-endpoint-to-use-fulfilment-date-calculator)

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ProductType.NewspaperVoucherBook
+  )
+
+  def apply(stage: Stage, fetchString: StringFromS3, today: LocalDate): Either[String, (ProductType, List[DayOfWeek]) => LocalDate] = {
+    for {
+      mappings <- productTypesWithFulfilmentDateFiles.traverse { productType =>
+        getStartDateForProductType(stage, fetchString, today, productType)
+          .map(startDate => productType -> startDate)
+      }
+      _ = logger.info(s"Successfully fetched start date mappings from fulfilment date files: ${mappings}")
```
Comment (kelvin-chappell) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala:37]: Perhaps move this line into yield block

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        .leftMap(ex => s"Failed to parse fulfilment file s3://$bucket/$key: $ex")
+      wireCatalog <- Either
+        .catchNonFatal(parseFulfillmentFile.as[Map[DayOfWeek, FulfilmentDates]])
+        .leftMap(ex => s"Failed to decode fulfilment file s3://$bucket/$key: $ex")
+      soonestDate = wireCatalog
+        .mapValues(_.newSubscriptionEarliestStartDate)
+        .collect { case (key, Some(value)) => key -> value }
+    } yield soonestDate
+  }
+}
```
Comment (kelvin-chappell) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala:77]: Would it be feasible to separate out pure and effectful code?  eg. get the catalogue from the bucket, and get the date from the catalogue

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/Catalog.scala [hunk 3]
```diff
@@ -211,11 +211,19 @@ case class WindowSizeDays(value: Int) extends AnyVal
...
 case class AmountMinorUnits(value: Int) extends AnyVal
 
+case class ProductType(value: String)
+object ProductType {
+  val GuardianWeekly = ProductType("Guardian Weekly")
+  val NewspaperVoucherBook = ProductType("Newspaper - Voucher Book")
+  val NewspaperHomeDelivery = ProductType("Newspaper - Home Delivery")
+  val DigitalPack = ProductType("Digital Pack")
+  val Contribution = ProductType("Contribution")
+}
```
Comment (kelvin-chappell) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/Catalog.scala:229]: Do you need to take the new digital vouchers into account, and the trial ones?

Comment (frj) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/Catalog.scala:235]: There are only fulfilment files for Weekly, Newspaper vouchers and Home Delivery so technically those are the only product types we need.

But for completeness Ive added the product types for all the products we actively use, including digital vouchers.

I think the trial was just using normal voucher subs so those subs would just have a product type of Newspaper - Voucher. Possibly something to bear in mind :)

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      mappings <- getStartDatesFromFulfillmentFiles(fulfilmentFileMap)
+    } yield {
+      logger.info(s"Successfully fetched start date mappings from fulfilment date files: ${mappings}")
+      lookupStartDateFunction(mappings) _
+    }
+  }
+
+  private def ascending(d1: LocalDate, d2: LocalDate) = d1.isBefore(d2)
+
+  def lookupStartDateFunction(startDateMappings: Map[ProductType, Map[DayOfWeek, LocalDate]])(productType: ProductType, issueDays: List[DayOfWeek]): LocalDate = {
```
Comment (kelvin-chappell) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/StartDateFromFulfilmentFiles.scala:44]:
```suggestion
  def lookupStartDate(startDateMappings: Map[ProductType, Map[DayOfWeek, LocalDate]])(productType: ProductType, issueDays: List[DayOfWeek]): LocalDate = {
```

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/NewProductApi.scala [hunk 5]
```diff
@@ -60,27 +74,29 @@ object NewProductApi {
...
+        startDate = today,
         maybeSize = Some(WindowSizeDays(1)),
         maybeCutOffDay = None,
         maybeStartDelay = None
-      ))
+      )
     )
 
     val windowOf90daysStartingIn2Weeks =  WindowRule(
+      startDate = today.plusDays(14),
```
[the remark this answers is not in the document]
Comment (frj) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/productcatalog/NewProductApi.scala:92]: Ill extract the numbers into constants. Bear in mind many of the constants in this file will go away as those field will be removed in the next PR

**Pull request merged**